### ISSUE TITLE
docs: Does module public API imply anything about package API?

### ISSUE THREAD

**Issue by nsajko** [JuliaLang/julia#51469]
Marking a name in a module with `public` makes the name part of the public API of that module. But what about the package that owns the module? How to define the package's API? It seems like module API may be conflated currently with package API?

An example: suppose I've got a package called *A*:
```julia
module A
  module B
    public c
    const c = 3.14
  end
end
```

**Is A.B.c part of the public API of the package?** IMO, it should not be (unless an additional `public B` declaration were added), but this isn't completely clear from the documentation.

**Comment by dalum** [comment]
I agree. This would follow my intuition about `export`, where it only exports the symbol one module out, and needs to be re-exported from the parent package module in order to be seen as properly public. Perhaps `public` declarations in an internal submodule wouldn't be immediately useful, but I can imagine tooling that could warn against using non-public symbols from such a submodule, as a way to ensure package integrity.

**Comment by vtjnash** [comment]
I think it already directly follows from the accessibility of it that `c` is a public export only from a private part of the package. So if you have the internals of A and thus access to `B`, then `c` is public to you. But if `B` is private to you, then `c` is not reachable.
